### PR TITLE
Datumizes customization styles

### PR DESCRIPTION
**Pull request by 112358sam**
[cleanliness]

## About the PR 
The global list of customization styles that include hair, facial hair, heterochromia, and makeup is removed and replaced with datumized alternatives. The /datum/customization_style abstract class is added with abstract subtypes for each category of hair. Changes are included that allow save files with the original strings to still work.

## Why's this needed? 
This will centralize everything related to customization styles to one file as currently there are several global lists to categorize the different styles. Some of these are not up to date with newer additions. This also lets the type of customization style be quickly checked by just checking the type rather than checking for membership by looping through a list.

### CODE REVIEW

**Comment by 112358sam** [comment]
Currently a draft because I expect there are issues and I want to give a few people a chance to look at it before submitting it. However, I've been testing a fair bit and everything seems to be working on local. I also expect there may be conflicts with the secret submodule.

**Comment by 112358sam** [comment]
Setting it ready for review now since I still haven't found any issues after a bit more testing.

**Comment by ZeWaka** [comment]
> I'm not sure how garbage collection works in BYOND, so I'm concerned some of my code will cause a memory leak.

@112358sam Here's something I wrote: https://spacestation13.github.io/DMByExample/objs/disposal.html

**Comment by 112358sam** [comment]
Thanks! That whole guide looks very useful. I'd expect my implementation should be okay then. The references to the datums are either restricted in scope to procs or limited to the three per human. The garbage collection should handle the de-referenced instances after they are overwritten.

**Comment by Tarmunora** [comment]
Hm, glancing through this, had a thought:
Why not tie customization color into the datum instead of keeping it as separate vars?

**Comment by 112358sam** [comment]
That's possible, but my concern is that changes to customization style and customization color are separate in practice. Hair color doesn't change when you get a haircut or use hairgrownium. So in most cases where hair style changes, it will need to copy the color of the previous style into the new one. I feel like it would complicate things more than simplify. However, I can go that way if you all would like.

File: code/datums/customization_styles.dm [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+		hetcroL
+			name = "Heterochromia: Left"
+			id = "hetcroL"
+		hetcroR
+			name = "Heterochromia: Right"
+			id = "hetcroR"
+
+proc/select_custom_style(list/customization_types, mob/living/carbon/human/user as mob)
+	var/list/options = list()
+	for (var/styletype as anything in customization_types)
```
Comment (ZeWaka) [inline code/datums/customization_styles.dm:575]: would like these and similar to be typed variables

File: code/datums/customization_styles.dm [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+ABSTRACT_TYPE(/datum/customization_style/sideburns)
+ABSTRACT_TYPE(/datum/customization_style/eyebrows)
+ABSTRACT_TYPE(/datum/customization_style/makeup)
+ABSTRACT_TYPE(/datum/customization_style/biological)
+
+/datum/customization_style
+	var/name = null
+	var/id = null
+	var/fem = 0
+	var/masc = 0
```
Comment (ZeWaka) [inline code/datums/customization_styles.dm:18]: could've been a single field utilizing bitflags but not entirely necessary to change

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/datums/customization_styles.dm (hunk 1 above)
Comment (112358sam) [reply code/datums/customization_styles.dm]: Thank you, I've updated the code to use more explicit typing.

File: code/datums/customization_styles.dm (hunk 2 above)
Comment (112358sam) [reply code/datums/customization_styles.dm]: Thank you, I've updated the implementation to use a single gender variable with bit flags.

File: code/datums/preferences.dm [hunk 3]
```diff
@@ -1705,137 +1706,33 @@ datum/preferences
...
+	var/datum/customization_style/style = new type
+	. = 0
+	if (style.gender & FEMININE)
+		. = 1
+
+proc/ismasc(type)
+	var/datum/customization_style/style = new type
+	. = 0
+	if (style.gender & MASCULINE)
+		. = 1
```
[the remark this answers is not in the document]
Comment (112358sam) [reply code/datums/preferences.dm]: Tested successfully

**Review by Tarmunora — APPROVED** [review]

**Pull request merged**